parsers/UY.py

```diff
@@ -1,163 +1,224 @@
+        return df_data
+    else:
+        raise ParserException(
+            parser="UY.py",
+            message="no data available for target_dateitme",
+        )
+
+
+def fix_solar_production(dt: datetime, row: pd.Series) -> int:
+    """sets solar production to 0 during the night as there is only solar PV in UY"""
+    if (5 >= dt.hour or dt.hour >= 20) and row.get("value") != 0:
+        return 0
+    else:
+        return round(row.get("value"), 3)
+
+
+@refetch_frequency(timedelta(days=1))
 def fetch_production(
     zone_key: str = "UY",
     session: Optional[Session] = None,
     target_datetime: Optional[datetime] = None,
     logger: Logger = getLogger(__name__),
-) -> dict:
-    if target_datetime:
-        raise NotImplementedError("This parser is not yet able to parse past dates")
-
-    obj = parse_page(session)
-
-    data = {
-        "zoneKey": zone_key,
-        "datetime": obj["datetime"],
-        "production": dict(
-            [(k, obj[INV_MAP_GENERATION[k]]) for k in INV_MAP_GENERATION.keys()]
-        ),
-        "source": "ute.com.uy",
-    }
-
-    return data
+) -> list:
+    """collects production data from ADME and format all data points for target_datetime"""
+    if target_datetime is None:
+        target_datetime = arrow.utcnow().datetime
+    session = session or Session()
 
+    data = fetch_data(
+        zone_key=zone_key,
+        session=session,
+        target_datetime=target_datetime,
+        sheet_name="GPF",
+        logger=logger,
+    )
+    data = data.rename(columns=PRODUCTION_MODE_MAPPING)
+    data = data.groupby(data.columns, axis=1).sum()
+    production = data[[col for col in data.columns if col in PRODUCTION_MODES]]
+    production = pd.melt(production, var_name="production_mode", ignore_index=False)
+    all_data_points = []
+    for dt in production.index.unique():
+        production_dict = {}
+        data_dt = production.loc[production.index == dt]
+        for i in range(len(data_dt)):
+            row = data_dt.iloc[i]
+            mode = row["production_mode"]
+            if mode == "solar":
+                production_dict[mode] = fix_solar_production(dt=dt, row=row)
+            else:
+                production_dict[mode] = round(row.get("value"), 3)
+        data_point = {
+            "zoneKey": "UY",
+            "datetime": arrow.get(dt).datetime.replace(tzinfo=pytz.timezone(UY_TZ)),
+            "production": production_dict,
+            "source": "pronos.adme.com.uy",
+        }
+        all_data_points += [data_point]
+    return all_data_points
+
+
+@refetch_frequency(timedelta(days=1))
+def fetch_consumption(
+    zone_key: str = "UY",
+    session: Optional[Session] = None,
+    target_datetime: Optional[datetime] = None,
+    logger: Logger = getLogger(__name__),
+) -> list:
+    """collects consumption data from ADME and format all data points for target_datetime"""
+    if target_datetime is None:
+        target_datetime = arrow.utcnow().datetime
```

Comment:
Shouldn't this be localized here? Because I guess their datasource is running on their timezone ?

Reply:
I didn't localize here because i feel like it's not something that is done in other parser but i may be completely wrong

Reply:
That is usually dependant on the source data. Most "good" sources are fully iso compliant and use UTC time for the timestamps but this may not be the case here (I have not looked myself).